### DISCUSSION TITLE
Error with nested write to table with composite primary key

### DISCUSSION THREAD

**Discussion by pradyuman** [prisma/prisma#4407] [Q&A] [answered]
I'm running into the following error when trying to do a nested write into a table with a composite primary key:
`Type '{ name: string; post: { connect: { id: string; }; }; }' is not assignable to type 'undefined'`.

Here's a sample schema:
```prisma
model User {
  id                                String                           @id @default(dbgenerated())
  name                              String
  organizationUsers                 OrganizationUser[]
  organizationUserPosts             OrganizationUserPost[]
}

model Organization {
  /// This field's default value can currently not be parsed: `uuid_generate_v4()`.
  id                                String                           @id @default(dbgenerated())
  name                              String
  organizationUsers                 OrganizationUser[]
  organizationUserPosts             OrganizationUserPost[]
}

model Post {
  id                                String                           @id @default(dbgenerated())
  organizationUserPosts             OrganizationUserPost[]
}

model OrganizationUser {
  organizationId                    String                           @map("organization_id")
  userId                            String                           @map("user_id")
  roles                             Json                             @default(dbgenerated())

  user                              User                             @relation(fields: [userId], references: [id])
  organization                      Organization                     @relation(fields: [organizationId], references: [id])
  organizationUserPosts             OrganizationUserPost[]

  @@id([organizationId, userId])
}

model OrganizationUserPost {
  organizationId     String              @map("organization_id")
  userId             String              @map("user_id")
  postId             String              @map("post_id")
  name               String

  user               User                @relation(fields: [userId], references: [id])
  post               Post                @relation(fields: [postId], references: [id])
  organization       Organization        @relation(fields: [organizationId], references: [id])
  organizationUser   OrganizationUser    @relation(fields: [organizationId, userId], references: [organizationId, userId])

  @@id([organizationId, userId, postId])
}
```

And a sample query:
```ts
  await prisma.organizationUser.create({
    data: {
      roles: ["owner"],
      organization: {
        create: {
          name: "Organization"
        }
      },
      user: {
        create: {
          name: "Entity",
        },
      },
      organizationUserPosts: {
        create: {
          name: "Post",
          post: {
            connect: {
              id: "938ba86e-794f-4fec-a76b-35271dd92662",
            },
          },
        },
      },
    },
    include: {
      organization: true,
      user: true,
    },
  })
```

What am I misunderstanding or doing incorrectly here?

**Comment by nikolasburk** [comment]
Hey @pradyuman, thanks a lot for raising this!

So, it seems like the `OrganizationUserPost` model has required 1-1-relationships to `Post` and `Organization` in your schema:

```prisma
model OrganizationUserPost {
  organizationId String  @map("organization_id")
  userId         String  @map("user_id")
  postId         String @map("post_id")
  name           String

  user         User         @relation(fields: [userId], references: [id])
  organization Organization @relation(fields: [organizationId], references: [id])

  // Post                           Post?             @relation(fields: [postId], references: [id])
  OrganizationUser               OrganizationUser? @relation(fields: [organizationUserOrganizationId, organizationUserUserId], references: [organizationId, userId])
  organizationUserOrganizationId String?
  organizationUserUserId         String?
  organizationUserPost           Post             @relation(fields: [postId], references: [id])

  @@id([organizationId, userId, postId])
}
```

Also note that your schema was missing the required virtual relation field e.g. called `organizationUserPost` in the schema. [Prisma always requires relation fields to be specified on both sides of relation](https://www.prisma.io/docs/concepts/components/prisma-schema/relations):

![image](https://user-images.githubusercontent.com/4058327/100591583-45d71800-32f6-11eb-9167-b1ff5772a11f.png)

However, in the code contained in `context.ts`, you're trying to create a record of type `OrganizationUserPost` that doesn't have connections to `Organization` nor to `Post`, but only to `OrganizationUser`, so you'll need to include that in your `create` query as well:

```diff
await prisma.organizationUser.create({
  data: {
    roles: ["owner"],
    organization: {
      create: {
        name: "Organization"
      }
    },
    user: {
      create: {
        name: "Entity",
      },
    },
    organizationUserPosts: {
      create: {
        name: "Post",
        user: {
          connect: {
            id: "938ba86e-794f-4fec-a76b-35271dd92662",
          },
        },
+        organization: {
+          connect: { id: ''}
+        },
+        organizationUserPost: {
+          connect: { id: ''}
+        }
      },
    },
  },
  include: {
    organization: true,
    user: true,
  },
})
```

Let me know if that solves your issue 🙌

**Reply by pradyuman** [reply]
For illustration purposes, the organization and post relations should be considered 1 -> many. I derived this from the full schema I sent you via DM.

I also just realized that the example I attached was an incorrect/outdated zip (sorry!). For clarity, I've updated the original post with the correct schema and query which illustrates the issue better.

I would expect that since I'm creating the "OrganizationUserPost" as a nested write, that Prisma would automatically be able to infer the organization and user ids required to write that row. However I still get a similar issue:
`Type '{ name: string; post: { connect: { id: string; }; }; }' is not assignable to type 'undefined'.`

**Reply by pradyuman** [reply]
If I understood your response correctly, what Prisma will do here is take my "connect" clause for the Post relation, infer OrganizationUser, but error because I haven't specified anything for the Organization or User relations. But this doesn't make sense to me, because what we care about to create the record are the organization and user ids, which are both specified in OrganizationUser and should carry over given that I'm creating OrganizationUserPost as a nested write within the OrganizationUser create call.

If this is a limitation of the prisma engine right now, I'm happy to make a GitHub issue to track this feature request, because my use case would be impossible without it.

In theory `uncheckedScalarInputs` might be able to fix this, but I'm running into a different set of issues related to that (and tracked in a different discussion).

**Reply by nikolasburk** [reply]
Ah ok, I _think_ I understand the issue now, thanks for the update!

So, the issue with the code above is that it's missing a link to an `Organization` and to a `User` inside of the `organizationUserPosts` nested writed:

```diff
await prisma.organizationUser.create({
  data: {
    roles: ["owner"],
    organization: {
      create: {
        name: "Organization"
      }
    },
    user: {
      create: {
        name: "Entity",
      },
    },
    organizationUserPosts: {
      create: {
        name: "Post",
        post: {
          connect: {
            id: "938ba86e-794f-4fec-a76b-35271dd92662",
          },
        },
+        organization: {
+          connect: { id: '' }
+        },
+        user: {
+          connect: { id: '' }
+        }
      },
    },
  },
  include: {
    organization: true,
    user: true,
  },
})
```

This will make the TS compiler happy, but I don't think this is what you want since you probably want to _connect_ the `User` and the `Organization` that you created on the first level of nesting – so you don't actually have the `id`s of the `User` and of the `Organization` in your code at this point because these are yet to be created?

Is that correct?

I'll try to figure out how you can achieve this, but I assume you probably need to create more than query for this 🤔

**Reply by nikolasburk** [reply]
Does this code do what you want?

```ts
  const user = await prisma.user.create({
    data: {
      name: 'Entity',
    },
  })
  const organization = await prisma.organization.create({
    data: {
      name: 'Organization',
    },
  })

  await prisma.organizationUser.create({
    data: {
      roles: ['owner'],
      organization: {
        connect: { id: organization.id },
      },
      user: {
        connect: { id: user.id },
      },
      organizationUserPosts: {
        create: {
          name: 'Post',
          post: {
            connect: {
              id: '938ba86e-794f-4fec-a76b-35271dd92662',
            },
          },
          organization: {
            connect: { id: organization.id },
          },
          user: {
            connect: { id: user.id },
          },
        },
      },
    },
    include: {
      organization: true,
      user: true,
    },
  })
```

**Reply by pradyuman** [reply]
As you guessed, I don't have the IDs yet so I was hoping that there was a way to "connect" them automatically. I haven't looked into the implementation, but there's a decent chance `uncheckedScalarInputs` can support this use case (if it doesn't already work out of the box).

My current workaround is to send two queries like you mentioned.

I'll track the GitHub Issue for the other bug and will create a new GitHub issue for this use case if it's still not possible to do this after the bugfix.

**Reply by nikolasburk** [reply]
> I haven't looked into the implementation, but there's a decent chance `uncheckedScalarInputs` can support this use case (if it doesn't already work out of the box).

Hmm maybe I'm missing something but I don't think being able to set an FK directly would change the situation since the core problem (being that you don't have access to the `id`s of the newly created `User` and `Organization` records) still would exist, no?  🤔

**Reply by nikolasburk** [reply]
But yeah, tracking this in GitHub issues is the way to go! In any case, thanks a lot for bringing this up 🙌

**Reply by pradyuman** [reply, marked as answer]
When I had `uncheckedScalarInputs` activated, I didn't get any type errors when compiling. That led me to believe that `uncheckedScalarInputs` might just be passing the foreign keys directly without validating if each relation was specified. Given that OrganizationUser's composite key is composed of the two ids that are required, if those are being passed in, the query may work.

This is predicated on my assumptions of how the Prisma engine works, which could be completely wrong. It's only been a few days since I started using it and haven't had the time or need to delve into the source code yet (so I don't actually know what's going on under the hood).

Thanks for your help! At least I learned that every relation needs to be specified for a nested write under the current implementation so I can take that into account going forward.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
